DDS-1098-container_deletion_parent

dmlond opened the pull request:
When a Folder or DataFile is deleted, it is moved to the root of the project trashbin, but its original parent is preserved for future restores.

dmlond committed:
adding deleted_from_parent_id attribute to containers to facilitate deletion into root of the project trashbin

dmlond committed:
DataFileSpec and FolderSpec expect DataFile and Folder to
  implement '#set_deleted_from_parent':
    if being deleted, set deleted_from_parent_id to parent_id and parent_id to nil
    if being restored, set parent_id to deleted_from_parent_id and deleted_from_parent_id to nil
  belong_to deleted_from_parent

dmlond committed:
FilesAPISpec expects FilesAPI to put folder child file into root of the project trashbin on delete
FoldersAPISpec expects FoldersAPI to put folder child folder into root of the project trashbin, but
  leave folder children in folder on delete

dmlond committed:
Merge branch 'DDS-829-trash_bin' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-1098-container_deletion_parent

dmann commented on app/models/container.rb:
```diff
@@ -37,4 +39,14 @@ def parent_id=(val)
     self.project = self.parent.project if self.parent
   end
+
+  def set_deleted_from_parent
```
I think it would be good to break this method into two methods. Something like `delete_to_trash` and `restore_to_parent`. They could encapsulate setting `is_deleted` as well as `deleted_from_parent_id` and `parent_id`. This should result in code that is easier to both read and test. 🤞

dmann reviewed (changes requested):
🍳

dmlond committed:
Merge branch 'DDS-829-trash_bin' into DDS-1098-container_deletion_parent

dmlond committed:
DataFile and Folder no longer implement set_deleted_from_parent
instead, implement
  - move_to_trashbin
    - sets is_deleted to true
    - sets deleted_from_parent_id to parent_id
    - sets parent_id to nil
  - restore_from_trashbin
    - default
     - sets is_deleted to false
     - sets parent_id to deleted_from_parent_id
     - sets deleted_from_parent_id to nil
    - new parent folder
     - sets is_deleted to false
     - sets parent_id to new_parent id
     - sets deleted_from_parent_id to nil
    - project
     - sets is_deleted to false
     - sets deleted_from_parent_id to nil
based on Code Review

dmlond committed:
FolderSpec and and ProjectSpec no longer test restore of a container from another project
as this is a validation error under the new design (project_id is immutable)

Project and Folder use restore_from_trashbin to restore the child

dmlond committed:
delete /files/:id and delete /folders/:id use .move_to_trashbin

dmlond committed:
added deleted_from_parent alias to parent to allow trashbin/dds-file-version/{id}/restore to access the parent data_file
in the same way it accesses object object parents

dmlond committed:
TrashbinApi /restore authorizes delete on purge_object

TrashbinAPISpec adds test coverage for all types of requests to /trashbin/{dds-kind}/{id}/restore

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
 
-    describe '#set_deleted_from_parent' do
-      it { is_expected.to respond_to(:set_deleted_from_parent) }
+    describe '.move_to_trashbin' do
```
`s/\./#/`
It's an instance method.

dmlond replied:
Sorry, I get these confused

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+    end
+  end
+
+  describe '.restore_from_trashbin' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+      it {
+        expect(subject.parent_id).not_to be_nil
+        expect(subject.parent).not_to be_nil
+        original_parent = subject.parent
```
You could set `original_parent` as a let variable and then, instead of having `expect(subject.parent).not_to be_nil`, use `expect(original_parent).not_to be_nil`

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+    context 'to original parent' do
+      it {
+        subject.move_to_trashbin
+        subject.save
```
Make sure the call to save is truthy.

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+        expect(subject.parent).to be_nil
+        expect(subject.deleted_from_parent_id).not_to be_nil
+        expect(subject.deleted_from_parent).not_to be_nil
+        original_parent = subject.deleted_from_parent
```
See above advice regarding `original_parent`

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+      let(:new_parent) { other_folder }
+      it {
+        subject.move_to_trashbin
+        subject.save
```
Make sure call to save is truthy

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+        expect(subject.parent).to be_nil
+        expect(subject.deleted_from_parent_id).not_to be_nil
+        expect(subject.deleted_from_parent).not_to be_nil
+        original_parent = subject.deleted_from_parent
```
See above advice regarding `original_parent`...

dmann replied:
Actually, if above 8 lines are the same for all these tests, then they belong in a before(:each) block as well. Just create a wrapping context called 'with trashed subject' or something.

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -262,53 +262,100 @@
+      }
+    end
+
+    context 'to project root' do
```
This tests how the method behaves when passed the subject's project. What happens with a different project?

dmann replied:
I'm also curious what happens when it's passes a folder from a different project.

dmlond replied:
The function does what it is told to do when a new project, or folder from a new project is passed. It sets these variables. This results in an invalid object based on the validations (project is immutable), which the trashbin_api restore method now communicates to the user.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -197,10 +197,11 @@
 
 
-  describe '#restore' do
+  describe '.restore' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -229,62 +230,40 @@
   end
 
-  describe '#purge' do
+  describe '.purge' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -313,26 +293,44 @@
 
-  describe '#set_deleted_from_parent' do
-    it { is_expected.to respond_to(:set_deleted_from_parent) }
+  describe '.move_to_trashbin' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -313,26 +293,44 @@
 
-    context 'being restored' do
-      subject { FactoryBot.create(:folder, :deleted, :with_parent) }
+  describe '.restore_from_trashbin' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/project_spec.rb:
```diff
@@ -372,7 +372,7 @@
     end #purge_children
 
-    describe '#restore' do
+    describe '.restore' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -196,10 +197,11 @@
       before do
-        subject.update_columns(is_deleted: true)
+        subject.move_to_trashbin
+        subject.save
```
Make sure the call to save is truthy

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -228,62 +230,40 @@
-        end
+      before do
+        child.move_to_trashbin
+        child.save
```
Make sure the call to save is truthy

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -298,7 +278,8 @@
       before do
-        subject.update_columns(is_deleted: true)
+        subject.move_to_trashbin
+        subject.save
```
Make sure the call to save is truthy

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -311,4 +292,100 @@
     end
   end
+
+  describe '.move_to_trashbin' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -311,4 +292,100 @@
+    }
+  end
+
+  describe '.restore_from_trashbin' do
```
`s/\./#/`
It's an instance method.

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -311,4 +292,100 @@
+    context 'to original parent' do
+      it {
+        subject.move_to_trashbin
+        subject.save
```
Make sure the call to save is truthy

dmann reviewed (changes requested):
I stopped reviewing about half way through. I found myself repeating my comments. I'll come back and review again tomorrow or Monday.

dmlond committed:
DataFileSpec expects restore_from_trashbin to throw exception if passed anything that is not a Folder of Project

DataFileSpec and FolderSpec refactored to use let for original_parent,
and target_project per Code Review Suggestion

dmlond committed:
expect .saves to be truthy per Code Review suggestion

dmlond committed:
added 'trashed resource' shared_context to models_shared_examples which
moves subject or trashed_resource_sym to the trashbin

DataFileSpec and FolderSpec use trashed resource context in #restore_from_trashbin tests

DataFileSpec and FolderSpec DRYed up by placing original_parent and trashed resource context into
base of '#restore_from_trashbin', along with a common before block to ensure that
the subject meets pre test expectations of parent, deleted_from_parent, and original_parent

dmlond committed:
.purge -> #purge based on Code Review suggestion

dmann commented on app/models/container.rb:
```diff
@@ -37,4 +39,27 @@ def parent_id=(val)
+    elsif new_parent.is_a? Folder
+      self.deleted_from_parent_id = nil
+      self.parent_id = new_parent.id
+      self.project_id = new_parent.project_id
```
The project automatically gets set to the parent's project via the `set_parent_attribute` callback (which calls `set_project_to_parent_project`). Should be able to get rid of this line and have your tests continue to pass.

dmann commented on app/models/container.rb:
```diff
@@ -37,4 +39,27 @@ def parent_id=(val)
+      self.project_id = new_parent.project_id
+    elsif new_parent.is_a? Project
+      self.deleted_from_parent_id = nil
+      self.project_id = new_parent.id
```
Should `self.parent` be set to `nil` here as well?

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -260,6 +262,138 @@
+      expect(original_parent).not_to be_nil
+    end
+
+    context 'to original parent' do
```
This is just one of two 'original parent' scenarios, where the original parent was a folder. The second is when the original parent was the project. I don't think it'll cause you to change the code, but it would be good to have both for test coverage.

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -260,6 +262,138 @@
+      end
+    end
+
+    context 'to original project root' do
```
This might be the second 'original parent' test I mentioned above, or at least pretty close to it. I think the key difference is to call `#restore_from_trashbin` without a target parameter.

dmann commented on spec/requests/trashbin_api_spec.rb:
```diff
@@ -64,83 +64,178 @@
+            it_behaves_like 'a client error' do
+              let(:expected_response) { 404 }
+              let(:expected_reason) { "dds-project #{project.id} is permenantly deleted, and cannot restore children." }
+              let(:expected_suggestion) { "Restore to a different project." }
```
I'm not asking for a change here, but is this even possible? I'm under the impression that when a project is deleted, it purges its children. Maybe this can happen with projects/children that are currently `is_deleted` in dds?

dmlond replied:
It may just be defensive

dmann commented on spec/requests/trashbin_api_spec.rb:
```diff
@@ -64,83 +64,178 @@
+            expect(parent_folder).to be_persisted
+          end
+
+          context 'that is deleted' do
```
I like this test. 👍

dmann commented on spec/requests/trashbin_api_spec.rb:
```diff
@@ -64,83 +64,178 @@
     end
 
     context 'file_version' do
+      subject { put(url, headers: headers) }
```
Why the addition of `subject { put(url, headers: headers) }` in this section?

dmann reviewed (changes requested):
A lot of good stuff. Just a few changes and questions. 🦆

dmlond committed:
removed redundant set of project_id. per Code Review
set parent_id to nil when new parent is a Project

dmlond committed:
added test coverage for restore_from_trashbin default given object originally in project root

dmlond committed:
begin using behaves_like a PUT request

dmlond committed:
further use of behaves_like a PUT request

dmlond committed:
further behaves_like a PUT request

dmlond committed:
further behaves_like a PUT request

dmlond committed:
fully refactored to use 'behaves like "A PUT request"

dmlond committed:
in PUT /trashbin/{kind}/{id}/purge:
  - added context to a naked 'a client error' block
  - made other contexts descriptions more appropriate
